Topic/macro def

retronym opened the pull request:
Submitting for review. It works, but needs some polish before it's ready to be merged.
- What are you after in Relations?
- Should we base scripted tests on 2.10.0-M2?

retronym committed:
Macro def aware recompilation.

 - Read macro modifier from method definition.
 - Always recompile downstream files after a file containing macro defs is recompiled.
 - Source is extended with a hasMacro attribute. Mark suggests that this might be better
   tracked in Relations, but I'm not sure how to make that change.

retronym committed:
whitespace

harrah commented:
With Relations, I was trying to think of how it would work when you use a macro defined in a binary.  With hasMacro being in Source, I don't think sbt will properly (for even a weak definition of properly) handle changes to macros in a binary dependency.

Yes, 2.10.0-M2 is good for tests.

retronym committed:
Avoid copy-paste of Flags.MACRO with source-compatibility trickery.

retronym committed:
Move condition out of SameAPI.

It was tenuously relying on the caller short-circuiting with `shortcutSameSource`.

retronym committed:
Scripted test for a macro def in a source file.

Needed to increase MaxPermSize, otherwise I was unable to run this test on Java 1.6.0_29 64bit Server on Mac OS.

retronym commented:
I can't see the problem with macros in JARs. Isn't SBT already suitably pessimistic in case of a modified binary?

```
[success] Total time: 3 s, completed Mar 4, 2012 11:13:48 PM
> compile
[success] Total time: 0 s, completed Mar 4, 2012 11:13:58 PM

> eval "touch /Users/jason/code/macrocosm/target/scala-2.10/macrocosm_2.10.0-SNAPSHOT-0.1-SNAPSHOT.jar" !
[info] ans: Int = 0

> compile
[info] Compiling 1 Scala source to /Users/jason/code/scratch/20120304/target/scala-2.9.1/classes...

> last compile
[debug] 
[debug] Initial source changes: 
[debug]     removed:Set()
[debug]     added: Set()
[debug]     modified: Set()
[debug] Removed products: Set()
[debug] Modified external sources: Set()
[debug] Modified binary dependencies: Set(/Users/jason/code/macrocosm/target/scala-2.10/macrocosm_2.10.0-SNAPSHOT-0.1-SNAPSHOT.jar)
[debug] Initial directly invalidated sources: Set()
[debug] 
[debug] Sources indirectly invalidated by:
[debug]     product: Set()
[debug]     binary dep: Set(/Users/jason/code/scratch/20120304/Client.scala)
[debug]     external source: Set()
[debug] Initially invalidated: Set(/Users/jason/code/scratch/20120304/Client.scala)

> eval "cat Client.scala" !
import com.github.retronym.macrocosm.Macrocosm._

object Client {
    tree(0)
}
```

harrah commented:
Yes, I think you are right about binaries.  The rest looks good, thanks!

(For future reference, I don't get notified when commits are added.  I'm only notified when there are comments.)
